pulse2percept/models/beyeler2019.py

```diff
@@ -378,16 +384,24 @@ def find_closest_axon(self, bundles, xret=None, yret=None):
         # Build a long list of all axon segments - their corresponding axon IDs
         # is given by `axon_idx` above:
         flat_bundles = np.concatenate(bundles)
-        # For every pixel on the grid, find the closest axon segment:
-        if self.engine == 'cython':
-            closest_seg = fast_find_closest_axon(flat_bundles,
-                                                 xret.ravel(),
-                                                 yret.ravel())
-        else:
-            closest_seg = [np.argmin((flat_bundles[:, 0] - x) ** 2 +
-                                     (flat_bundles[:, 1] - y) ** 2)
-                           for x, y in zip(xret.ravel(),
-                                           yret.ravel())]
+        if self.use_legacy_build:
+            # For every pixel on the grid, find the closest axon segment:
+            if self.engine == 'cython':
+                closest_seg = fast_find_closest_axon(flat_bundles,
+                                                     xret.ravel(),
+                                                     yret.ravel())
+            else:
+                closest_seg = [np.argmin((flat_bundles[:, 0] - x) ** 2 +
+                                         (flat_bundles[:, 1] - y) ** 2)
+                               for x, y in zip(xret.ravel(),
+                                               yret.ravel())]
+        else: 
+            kdtree = cKDTree(flat_bundles, leafsize=60)
```

Comment:
Can you add a comment about how you chose `leafsize` - do you think this value will always work, or should it be something that can be set by the user?

Reply:
Benchmarking the best leaf size was somewhat difficult. I experimented with leaf sizes in the range [10, 300] for number of axons in the range [500, 10000], and eventually concluded that leaf size didn't affect the result much. 

When changing leaf size for a fixed number of axons, the total fit+query time never changed by more than 200ms (max - min < 200). Further when running multiple trials of the same leaf size and num_axons, the total time changed by up to ~170ms (max-min). Thus it was hard to determine if changing leaf size was actually affecting run time, or if it was just random variance between trials. 

After running a large number of trials, it seemed like any leafsize in the range [20, 100] was essentially identical, with marginal performance drops otherwise.

In general, the optimal leaf size increased slightly with number of axons, but since changing leaf size never affected the result by more than 200ms, I didn't think it would be worth spending a lot of time figuring out what the best leaf size was for a passed number of axons, and just left it as 50.

Reply:
So in theory the current value of leafsize should work pretty well.

Letting the user change it is extremely easy though, so I can also just add it as a parameter that can optionally be passed to find_closest_axon

Reply:
That's ok. We don't need to make it a settable option if the behavior doesn't change much. Choosing a good value for the user simplifies things. I was just wondering if it would make a difference.

Reply:
We can merge this as is. Great job!